File: stick-rumble-client/src/game/network/NetworkSimulator.ts
/**
 * NetworkSimulator - Simulates artificial network latency, jitter, and packet loss
 * for testing netcode under poor network conditions.
 */

export interface NetworkSimulatorConfig {
  latency?: number; // Base latency in ms (0-300)
  packetLoss?: number; // Packet loss percentage (0-20)
  enabled?: boolean; // Whether simulation is active
}

export interface NetworkSimulatorStats {
  enabled: boolean;
  latency: number;
  packetLoss: number;
}

export class NetworkSimulator {
  private latency = 0; // Base latency in ms
  private packetLoss = 0; // Packet loss percentage (0-100)
  private enabled = false;

  constructor(config?: NetworkSimulatorConfig) {
    if (config) {
      this.setLatency(config.latency ?? 0);
      this.setPacketLoss(config.packetLoss ?? 0);
      this.enabled = config.enabled ?? false;
    }
  }

  /**
   * Enable or disable network simulation
   */
  setEnabled(enabled: boolean): void {
    this.enabled = enabled;
  }

  /**
   * Check if network simulation is enabled
   */
  isEnabled(): boolean {
    return this.enabled;
  }

  /**
   * Set base latency in milliseconds (clamped to 0-300ms)
   */
  setLatency(latency: number): void {
    this.latency = Math.max(0, Math.min(300, latency));
  }

  /**
   * Get current base latency
   */
  getLatency(): number {
    return this.latency;
  }

  /**
   * Set packet loss percentage (clamped to 0-20%)
   */
  setPacketLoss(packetLoss: number): void {
    this.packetLoss = Math.max(0, Math.min(20, packetLoss));
  }

  /**
   * Get current packet loss percentage
   */
  getPacketLoss(): number {
    return this.packetLoss;
  }

  /**
   * Get current simulator stats
   */
  getStats(): NetworkSimulatorStats {
    return {
      enabled: this.enabled,
      latency: this.latency,
      packetLoss: this.packetLoss,
    };
  }

  /**
   * Simulate sending a message with artificial latency and packet loss
   * @param message - Message to send
   * @param sendFn - Actual send function to call after simulation
   */
  simulateSend<T>(message: T, sendFn: (msg: T) => void): void {
    if (!this.enabled || this.latency === 0) {
      // No simulation or zero latency - send immediately
      sendFn(message);
      return;
    }

    // Check for packet loss
    if (this.shouldDropPacket()) {
      // Packet dropped - don't send
      return;
    }

    // Calculate delay with jitter
    const delay = this.calculateDelay();

    // Delay the send
    setTimeout(() => {
      sendFn(message);
    }, delay);
  }

  /**
   * Simulate receiving a message with artificial latency and packet loss
   * @param message - Message received
   * @param receiveFn - Function to call after simulated delay
   */
  simulateReceive<T>(message: T, receiveFn: (msg: T) => void): void {
    if (!this.enabled || this.latency === 0) {
      // No simulation or zero latency - receive immediately
      receiveFn(message);
      return;
    }

    // Check for packet loss
    if (this.shouldDropPacket()) {
      // Packet dropped - don't receive
      return;
    }

    // Calculate delay with jitter
    const delay = this.calculateDelay();

    // Delay the receive
    setTimeout(() => {
      receiveFn(message);
    }, delay);
  }

  /**
   * Calculate delay with jitter (+/-20ms)
   */
  private calculateDelay(): number {
    const jitter = Math.random() * 40 - 20; // +/-20ms
    return Math.max(0, this.latency + jitter);
  }

  /**
   * Determine if packet should be dropped based on packet loss rate
   */
  private shouldDropPacket(): boolean {
    if (this.packetLoss === 0) {
      return false;
    }
    return Math.random() * 100 < this.packetLoss;
  }
}

File: stick-rumble-server/internal/network/network_simulator.go
package network

import (
	"log"
	"math/rand"
	"os"
	"strconv"
	"time"
)

// NetworkSimulator simulates artificial network latency and packet loss for testing
type NetworkSimulator struct {
	latency    int  // Base latency in milliseconds (0-300)
	packetLoss int  // Packet loss percentage (0-20)
	enabled    bool // Whether simulation is active
}

// NewNetworkSimulator creates a new network simulator from environment variables.
// Reads SIMULATE_LATENCY and SIMULATE_PACKET_LOSS environment variables.
// Returns nil if neither variable is set.
func NewNetworkSimulator() *NetworkSimulator {
	latencyStr := os.Getenv("SIMULATE_LATENCY")
	packetLossStr := os.Getenv("SIMULATE_PACKET_LOSS")

	// If neither env var is set, return nil (no simulation)
	if latencyStr == "" && packetLossStr == "" {
		return nil
	}

	sim := &NetworkSimulator{
		enabled: true,
	}

	// Parse latency
	if latencyStr != "" {
		latency, err := strconv.Atoi(latencyStr)
		if err != nil {
			log.Printf("[NetworkSimulator] Invalid SIMULATE_LATENCY value: %s", latencyStr)
		} else {
			sim.SetLatency(latency)
		}
	}

	// Parse packet loss
	if packetLossStr != "" {
		packetLoss, err := strconv.Atoi(packetLossStr)
		if err != nil {
			log.Printf("[NetworkSimulator] Invalid SIMULATE_PACKET_LOSS value: %s", packetLossStr)
		} else {
			sim.SetPacketLoss(packetLoss)
		}
	}

	if sim.latency > 0 || sim.packetLoss > 0 {
		log.Printf("[NetworkSimulator] Enabled with latency=%dms, packetLoss=%d%%", sim.latency, sim.packetLoss)
	}

	return sim
}

// SetLatency sets the base latency in milliseconds (clamped to 0-300ms)
func (s *NetworkSimulator) SetLatency(latency int) {
	if latency < 0 {
		latency = 0
	} else if latency > 300 {
		latency = 300
	}
	s.latency = latency
}

// GetLatency returns the current base latency
func (s *NetworkSimulator) GetLatency() int {
	return s.latency
}

// SetPacketLoss sets the packet loss percentage (clamped to 0-20%)
func (s *NetworkSimulator) SetPacketLoss(packetLoss int) {
	if packetLoss < 0 {
		packetLoss = 0
	} else if packetLoss > 20 {
		packetLoss = 20
	}
	s.packetLoss = packetLoss
}

// GetPacketLoss returns the current packet loss percentage
func (s *NetworkSimulator) GetPacketLoss() int {
	return s.packetLoss
}

// IsEnabled returns whether the simulator is enabled
func (s *NetworkSimulator) IsEnabled() bool {
	return s != nil && s.enabled
}

// ShouldDropPacket determines if a packet should be dropped based on packet loss rate
func (s *NetworkSimulator) ShouldDropPacket() bool {
	if s == nil || !s.enabled || s.packetLoss == 0 {
		return false
	}
	return rand.Intn(100) < s.packetLoss
}

// GetDelay calculates the delay to apply including jitter (+/-20ms)
func (s *NetworkSimulator) GetDelay() time.Duration {
	if s == nil || !s.enabled || s.latency == 0 {
		return 0
	}
	// Add jitter: +/-20ms
	jitter := rand.Intn(41) - 20 // -20 to +20
	delay := s.latency + jitter
	if delay < 0 {
		delay = 0
	}
	return time.Duration(delay) * time.Millisecond
}

// SimulateSend wraps a send function with artificial latency and packet loss.
// If the packet should be dropped, sendFn is not called.
// Otherwise, sendFn is called after the simulated delay.
func (s *NetworkSimulator) SimulateSend(sendFn func()) {
	if s == nil || !s.enabled {
		sendFn()
		return
	}

	// Check for packet loss first
	if s.ShouldDropPacket() {
		// Packet dropped - don't send
		return
	}

	// Get delay
	delay := s.GetDelay()
	if delay == 0 {
		sendFn()
		return
	}

	// Delay the send (asynchronously to not block)
	go func() {
		time.Sleep(delay)
		sendFn()
	}()
}
